Sources/NIO/ByteCollectionUtils.swift

```diff
@@ -0,0 +1,90 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+fileprivate let defaultWhitespaces = [" ", "\t"].map({UInt8($0.utf8CString[0])})
+
+public extension StringProtocol {
+    /// Gets the string as a contiguous array to be used with comparator methods.
+    /// Useful ONLY for constants or long lived strings.
+    public var asContiguousUTF8UIntArray: ContiguousArray<UInt8> {
+        return ContiguousArray.init(self.utf8)
+    }
+    
+    /// Gets the string as a contiguous array to be used with comparator methods.
+    /// But in Upper Case, so could use in case sensitive comparison, it already uppercases the
+    /// buffer.
+    public var asUpperCaseContiguousUTF8UIntArray: ContiguousArray<UInt8> {
+        return ContiguousArray.init(self.utf8.map({$0 & 0xdf}))
+    }
+
+}
+
+// Collection comparator extension
+// Could be used on any collection to compare their contents with any other one
+extension Collection where Element: Equatable {
+    func compare<T: Collection>(rhs: T, selfTransform: (T.Element) -> (T.Element) = { return $0 }) -> Bool where T.Element == Self.Element {
+        
+        // If lengths are not equal, it can't be equal
+        if self.count != rhs.count { return false }
+        
+        // Create an iterator for the right hand side, so that both are iterated, no need for indices
+        var rightIterator = rhs.makeIterator()
+        
+        for leftByte in self {
+            if (rightIterator.next()) != selfTransform(leftByte) {
+                return false
+            }
+        }
+        return true
+    }
+    
+    public static func ==<T: Collection>(lhs: Self, rhs: T) -> Bool where T.Element == Self.Element {
+        return lhs.compare(rhs: rhs)
+    }
+    
+    public static func !=<T: Collection>(lhs: Self, rhs: T) -> Bool where T.Element == Self.Element {
+        return !lhs.compare(rhs: rhs)
+    }
+}
+
+extension Collection where Self.Element: Equatable {
+    
+    public func trim(limitingElements: [Self.Element]) -> Self.SubSequence {
+        let trimmingLast = self.dropLast(self.reversed().enumerated().first(where: {!limitingElements.contains($0.element)})?.offset ?? 0)
```

Comment:
Mind splitting this up into two statements? It's sufficiently long that its purpose isn't very clear.